Command line completions

TAswan opened the pull request:
had a look at adding command line completions as discussed in #789 and found quite a few things to note:

- it doesn't look like being able to run this directly from build is possible like the documentation shows, as CLI uses the solve module, so I believe this would make conjure oxide a build dependency of conjure oxide. I think we could use cargo x-tasks for this (https://github.com/matklad/cargo-xtask) to automatically make this after building, but with its own limitations.

- most documentation talks about the non derive style of clap, so when it refers to building your cli, we don't have access to this. Instead we can use `CLI::command()` which is included in the parser derive.

- On my shell (Zsh) the completion file produced with our cli appears to be broken somewhat, but I haven't been able to pin down what functionality is and isn't currently broken, nor have I tested it with any other shells to see if they have more compatibility 

- additionally we can add some hints in the cli to hopefully make the completions more useful, eg hinting that something should be a filepath for target file (see a more comprehensive list here: https://github.com/clap-rs/clap/blob/master/clap_complete/examples/completion-derive.rs)

If you want to try it out then the current code generates at run time provided a stdout eg `./conjure_oxide help > test.zsh` (running through cargo should also work, but a command for oxide is required to stop early quitting) this should produce a message that it's generating a file for your shell (see list in code) and it should produce the file in the same directory as the compiled conjure oxide file.

 from there refresh your completions using `compinit` (varies from shell to shell, works for zsh and should work for bash) then check the completions by running ./conjure_oxide \<tab>  (this does have to be the compiled version, not cargo)

niklasdewally reviewed (commented):
> it doesn't look like being able to run this directly from build is possible like the documentation shows, as CLI uses the solve module, so I believe this would make conjure oxide a build dependency of conjure oxide. I think we could use cargo x-tasks for this (https://github.com/matklad/cargo-xtask) to automatically make this after building, but with its own limitations.

At runtime is fine.

My ideal workflow is like `pandoc`: in my `.bashrc` I have `eval $(pandoc --bash-completion)`, which loads pandoc's completion into my shell.

If we needed to distribute the completion as a standalone script (e.g. for homebrew), could always get CI to run the latest executable and save its completion output.

> additionally we can add some hints in the cli to hopefully make the completions more useful, eg hinting that something should be a filepath for target file (see a more comprehensive list here: https://github.com/clap-rs/clap/blob/master/clap_complete/examples/completion-derive.rs)

good idea :)

> If you want to try it out then the current code generates at run time provided a stdout eg ./conjure_oxide help > test.zsh (running through cargo should also work, but a command for oxide is required to stop early quitting) this should produce a message that it's generating a file for your shell (see list in code) and it should produce the file in the same directory as the compiled conjure oxide file.
> 
> from there refresh your completions using compinit (varies from shell to shell, works for zsh and should work for bash) then check the completions by running ./conjure_oxide <tab> (this does have to be the compiled version, not cargo)

I can't seem to get this to work: `./target/release/conjure_oxide help` gives me no script as output, just the help text. I don't get `unrecognised shell` either...

 I run homebrew `bash` on Mac (`SHELL=/opt/homebrew/bin/bash`)

I suggest hiding completion behind its own subcommand (e.g. to get bash completion, do `eval $(conjure_oxide completion --bash)`. E.g. the clap example you linked has a `--generate` flag to do this, and pandoc has `--bash-completion`.

TAswan commented:
Completions are now added and working! (with some caveats)

- implemented as a subcommand that prints the completion file to standard out (eg. conjure_oxide completion zsh > out.zsh)

- supports zsh well, some drawbacks on bash. primarily with it incorrectly suggesting <INPUT ESSENCE> rather than an input file. This appears to be a drawback of either the clap_complete library, or bash itself as value hints only work for fish and zsh (although from my usage zsh is very good at picking up on what should be files without hints). @niklasdewally if you find any other issues with regards to bash usage please let me know!

- should also support elvish, fish, and powershell, but I currently don't have access to any of those shells so if anyone uses them please let me know how it works for you!

typical usage on zsh would be to run `eval "$(cargo run completion zsh)"` then running via the conjure_oxide binary should allow for completions, some screenshots of usage below:

![image](https://github.com/user-attachments/assets/b540bb75-5fb9-4f93-a7e1-a0b647a2237a)

![image](https://github.com/user-attachments/assets/830ae7e4-176d-4b93-a763-40ca8f47c0f8)

![image](https://github.com/user-attachments/assets/7c15cdb2-97f7-46b7-b6e7-a0164b18fcfd)

ozgurakgun commented on conjure_oxide/src/main.rs:
```diff
@@ -125,13 +127,33 @@ fn setup_logging(global_args: &GlobalArgs) -> anyhow::Result<()> {
+    let shell = completion_args.shell;
+    let name = cmd.get_name().to_string();
+
+    let shell = match shell {
```
Wondering if we could use the existing `Shell` type directly instead of defining our own? Seems identical: https://docs.rs/clap_complete/latest/clap_complete/aot/enum.Shell.html
